File: bootcamp/semana-06/3-proyecto/solucion-proyecto-06.js
/**
 * ============================================
 * Solución: Proyecto Semana 06
 * Buscador de Texto Avanzado
 * ============================================
 */

// ============================================
// CONFIGURACIÓN
// ============================================

const SearchConfig = {
  caseSensitive: false,
  wholeWord: false,
  multiline: true,
  dotAll: false,
  unicode: true,
  showIndices: true,
};

// ============================================
// UTILIDADES
// ============================================

/**
 * Escapar caracteres especiales de regex
 *
 * ¿Por qué? Los caracteres como . * + tienen significado especial
 * ¿Para qué? Buscar literalmente el texto del usuario
 */
function escapeRegExp(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Construir string de flags según configuración
 *
 * @param {object} config - Configuración de búsqueda
 * @returns {string} Flags para regex
 */
function buildFlags(config) {
  let flags = 'g'; // Siempre global

  if (!config.caseSensitive) flags += 'i';
  if (config.multiline) flags += 'm';
  if (config.dotAll) flags += 's';
  if (config.unicode) flags += 'u';
  if (config.showIndices) flags += 'd';

  return flags;
}

/**
 * Construir patrón regex
 *
 * @param {string} query - Término de búsqueda
 * @param {object} config - Configuración
 * @returns {RegExp} Expresión regular
 */
function buildPattern(query, config) {
  let pattern = escapeRegExp(query);

  if (config.wholeWord) {
    // \b para límites de palabra
    pattern = `\\b${pattern}\\b`;
  }

  const flags = buildFlags(config);

  return new RegExp(pattern, flags);
}

// ============================================
// MOTOR DE BÚSQUEDA
// ============================================

/**
 * Motor de búsqueda principal
 *
 * ¿Por qué? Centralizar toda la lógica de búsqueda
 * ¿Para qué? Reutilización y mantenibilidad
 *
 * @param {string} texto - Texto donde buscar
 * @param {string} query - Término de búsqueda
 * @param {object} opciones - Configuración
 * @returns {object} Resultados
 */
function buscar(texto, query, opciones = {}) {
  const config = { ...SearchConfig, ...opciones };

  if (!query || !texto) {
    return {
      matches: [],
      total: 0,
      positions: [],
      highlighted: texto || '',
      stats: { duration: 0 },
    };
  }

  const startTime = performance.now();

  try {
    const pattern = buildPattern(query, config);
    const matches = [];
    const positions = [];

    // Usar matchAll para obtener toda la info
    for (const match of texto.matchAll(pattern)) {
      matches.push(match[0]);

      if (config.showIndices && match.indices) {
        positions.push({
          start: match.indices[0][0],
          end: match.indices[0][1],
          value: match[0],
        });
      } else {
        positions.push({
          start: match.index,
          end: match.index + match[0].length,
          value: match[0],
        });
      }
    }

    const highlighted = highlight(texto, pattern);
    const duration = performance.now() - startTime;

    return {
      matches,
      total: matches.length,
      positions,
      highlighted,
      stats: {
        duration: duration.toFixed(2) + 'ms',
        textLength: texto.length,
        pattern: pattern.toString(),
      },
    };
  } catch (error) {
    return {
      matches: [],
      total: 0,
      positions: [],
      highlighted: texto,
      stats: { error: error.message },
    };
  }
}

/**
 * Resaltar coincidencias en el texto
 *
 * @param {string} texto - Texto original
 * @param {RegExp} pattern - Patrón de búsqueda
 * @returns {string} Texto con marcadores
 */
function highlight(texto, pattern) {
  return texto.replace(pattern, '[[MATCH]]$&[[/MATCH]]');
}

/**
 * Convertir marcadores a HTML
 */
function highlightToHTML(texto) {
  return texto
    .replace(/\[\[MATCH\]\]/g, '<mark class="highlight">')
    .replace(/\[\[\/MATCH\]\]/g, '</mark>');
}

// ============================================
// BÚSQUEDA AVANZADA
// ============================================

/**
 * Parsear query con sintaxis avanzada
 *
 * Soporta:
 * - "frase exacta"
 * - palabra1 palabra2 (AND implícito)
 * - palabra1 OR palabra2
 * - -excluir
 * - palabra* (wildcard)
 *
 * @param {string} queryString - Query del usuario
 * @returns {object} Componentes parseados
 */
function parseQuery(queryString) {
  const parts = {
    exact: [],
    include: [],
    exclude: [],
    wildcard: [],
    or: [],
  };

  let remaining = queryString;

  // 1. Extraer frases exactas "..."
  const exactPattern = /"([^"]+)"/g;
  for (const match of remaining.matchAll(exactPattern)) {
    parts.exact.push(match[1]);
  }
  remaining = remaining.replace(exactPattern, ' ');

  // 2. Extraer exclusiones -palabra
  const excludePattern = /-(\w+)/g;
  for (const match of remaining.matchAll(excludePattern)) {
    parts.exclude.push(match[1]);
  }
  remaining = remaining.replace(excludePattern, ' ');

  // 3. Extraer wildcards palabra*
  const wildcardPattern = /(\w+)\*/g;
  for (const match of remaining.matchAll(wildcardPattern)) {
    parts.wildcard.push(match[1]);
  }
  remaining = remaining.replace(wildcardPattern, ' ');

  // 4. Procesar OR
  const orPattern = /(\w+)\s+OR\s+(\w+)/gi;
  for (const match of remaining.matchAll(orPattern)) {
    parts.or.push([match[1], match[2]]);
  }
  remaining = remaining.replace(orPattern, ' ');

  // 5. El resto son palabras a incluir (AND)
  const words = remaining.trim().split(/\s+/).filter(Boolean);
  parts.include.push(...words);

  return parts;
}

/**
 * Búsqueda avanzada con sintaxis especial
 *
 * @param {string} texto - Texto donde buscar
 * @param {string} queryString - Query con sintaxis
 * @param {object} opciones - Configuración base
 * @returns {object} Resultados
 */
function busquedaAvanzada(texto, queryString, opciones = {}) {
  const config = { ...SearchConfig, ...opciones };
  const parsed = parseQuery(queryString);
  const resultados = [];
  let textoFiltrado = texto;

  // Verificar exclusiones primero
  for (const excluir of parsed.exclude) {
    const pattern = new RegExp(`\\b${escapeRegExp(excluir)}\\b`, 'gi');
    if (pattern.test(texto)) {
      // Si contiene palabra excluida, podría filtrar líneas
      const lineas = texto.split('\n');
      textoFiltrado = lineas.filter((linea) => !pattern.test(linea)).join('\n');
    }
  }

  // Buscar frases exactas
  for (const frase of parsed.exact) {
    const resultado = buscar(textoFiltrado, frase, config);
    if (resultado.total > 0) {
      resultados.push({
        type: 'exact',
        query: frase,
        ...resultado,
      });
    }
  }

  // Buscar wildcards
  for (const base of parsed.wildcard) {
    const pattern = new RegExp(`\\b${escapeRegExp(base)}\\w*\\b`, 'gi');
    const matches = textoFiltrado.match(pattern) || [];
    resultados.push({
      type: 'wildcard',
      query: base + '*',
      matches: [...new Set(matches)],
      total: matches.length,
    });
  }

  // Buscar OR
  for (const [a, b] of parsed.or) {
    const pattern = new RegExp(
      `\\b(?:${escapeRegExp(a)}|${escapeRegExp(b)})\\b`,
      'gi'
    );
    const matches = textoFiltrado.match(pattern) || [];
    resultados.push({
      type: 'or',
      query: `${a} OR ${b}`,
      matches,
      total: matches.length,
    });
  }

  // Buscar palabras individuales (AND)
  for (const palabra of parsed.include) {
    const resultado = buscar(textoFiltrado, palabra, {
      ...config,
      wholeWord: true,
    });
    if (resultado.total > 0) {
      resultados.push({
        type: 'include',
        query: palabra,
        ...resultado,
      });
    }
  }

  // Verificar que todas las palabras include estén presentes (AND)
  const todasPresentes = parsed.include.every((palabra) => {
    const pattern = new RegExp(`\\b${escapeRegExp(palabra)}\\b`, 'gi');
    return pattern.test(textoFiltrado);
  });

  return {
    parsed,
    resultados,
    todasPresentes,
    textoFiltrado,
  };
}

// ============================================
// UTILIDADES ADICIONALES
// ============================================

/**
 * Obtener contexto alrededor de un match
 *
 * @param {string} texto - Texto completo
 * @param {number} posicion - Posición del match
 * @param {number} caracteres - Caracteres de contexto
 * @returns {string} Fragmento con contexto
 */
function getContexto(texto, posicion, caracteres = 50) {
  const start = Math.max(0, posicion - caracteres);
  const end = Math.min(texto.length, posicion + caracteres);

  let contexto = texto.slice(start, end);

  if (start > 0) contexto = '...' + contexto;
  if (end < texto.length) contexto = contexto + '...';

  return contexto;
}

/**
 * Obtener líneas alrededor de un match
 *
 * @param {string} texto - Texto completo
 * @param {number} posicion - Posición del match
 * @param {number} lineas - Líneas de contexto
 * @returns {object} Líneas con contexto
 */
function getLineasContexto(texto, posicion, lineas = 2) {
  const todasLineas = texto.split('\n');

  // Encontrar número de línea
  let charCount = 0;
  let lineaActual = 0;

  for (let i = 0; i < todasLineas.length; i++) {
    charCount += todasLineas[i].length + 1; // +1 por \n
    if (charCount > posicion) {
      lineaActual = i;
      break;
    }
  }

  const inicio = Math.max(0, lineaActual - lineas);
  const fin = Math.min(todasLineas.length, lineaActual + lineas + 1);

  return {
    lineas: todasLineas.slice(inicio, fin),
    numeroLineaMatch: lineaActual,
    rango: [inicio, fin],
  };
}

// ============================================
// TESTS
// ============================================

const textoEjemplo = `JavaScript es un lenguaje de programación.
JavaScript se usa para desarrollo web.
También se usa en Node.js para backend.
Python es otro lenguaje popular.
javascript puede ejecutarse en el navegador.
Los frameworks como React usan JavaScript.`;

console.log('=== Búsqueda Simple ===\n');

const resultado1 = buscar(textoEjemplo, 'JavaScript');
console.log("Buscar 'JavaScript':", resultado1.total, 'coincidencias');
console.log('Matches:', resultado1.matches);

console.log('\n=== Búsqueda Case-Sensitive ===\n');

const resultado2 = buscar(textoEjemplo, 'JavaScript', { caseSensitive: true });
console.log("Buscar 'JavaScript' (case-sensitive):", resultado2.total);

const resultado3 = buscar(textoEjemplo, 'javascript', { caseSensitive: true });
console.log("Buscar 'javascript' (case-sensitive):", resultado3.total);

console.log('\n=== Búsqueda Palabra Completa ===\n');

const resultado4 = buscar(textoEjemplo, 'Java', { wholeWord: true });
console.log("Buscar 'Java' (whole word):", resultado4.total);

const resultado5 = buscar(textoEjemplo, 'Java', { wholeWord: false });
console.log("Buscar 'Java' (parcial):", resultado5.total);

console.log('\n=== Búsqueda Avanzada ===\n');

const query1 = '"desarrollo web" JavaScript -Python';
console.log(`Query: ${query1}`);
console.log(busquedaAvanzada(textoEjemplo, query1));

console.log('\n=== Contexto de Matches ===\n');

const resultado6 = buscar(textoEjemplo, 'Node.js');
if (resultado6.positions.length > 0) {
  const pos = resultado6.positions[0].start;
  console.log('Contexto:', getContexto(textoEjemplo, pos, 30));
  console.log('Líneas:', getLineasContexto(textoEjemplo, pos, 1));
}

console.log('\n=== Unicode ===\n');

const textoUnicode = 'Hola 👋 mundo! Привет мир! 你好世界!';
const resultado7 = buscar(textoUnicode, 'мир', { unicode: true });
console.log("Buscar 'мир' en texto Unicode:", resultado7);

File: bootcamp/week-06-flags_y_modificadores/3-proyecto/solucion-proyecto-06.py
"""
============================================
Solución: Proyecto Semana 06
Buscador de Texto Avanzado
============================================
"""

import re
import time

# ============================================
# CONFIGURACIÓN
# ============================================

SEARCH_CONFIG = {
    'case_sensitive': False,
    'whole_word': False,
    'multiline': True,
    'dot_all': False,
    'unicode': True,
    'show_indices': True,
}

# ============================================
# UTILIDADES
# ============================================

def escape_regex(string: str) -> str:
    """
    Escapar caracteres especiales de regex

    ¿Por qué? Los caracteres como . * + tienen significado especial
    ¿Para qué? Buscar literalmente el texto del usuario
    """
    return re.sub(r'[.*+?^${}()|[\]\\]', r'\\\g<0>', string)


def build_flags(config: dict) -> int:
    """
    Construir bitmask de flags según configuración

    :param config: Configuración de búsqueda
    :returns: Flags para regex como entero bitmask
    """
    flags = 0  # No existe flag 'g' en Python; se usa findall/finditer en su lugar

    if not config['case_sensitive']:
        flags |= re.IGNORECASE
    if config['multiline']:
        flags |= re.MULTILINE
    if config['dot_all']:
        flags |= re.DOTALL
    if config['unicode']:
        flags |= re.UNICODE
    if config['show_indices']:
        # Python siempre expone .start() / .end() — no se necesita flag 'd'
        pass

    return flags


def build_pattern(query: str, config: dict) -> re.Pattern:
    """
    Construir patrón regex

    :param query: Término de búsqueda
    :param config: Configuración
    :returns: Expresión regular compilada
    """
    pattern = escape_regex(query)

    if config['whole_word']:
        # \b para límites de palabra
        pattern = rf'\b{pattern}\b'

    flags = build_flags(config)

    return re.compile(pattern, flags)


# ============================================
# MOTOR DE BÚSQUEDA
# ============================================

def buscar(texto: str, query: str, opciones: dict | None = None) -> dict:
    """
    Motor de búsqueda principal

    ¿Por qué? Centralizar toda la lógica de búsqueda
    ¿Para qué? Reutilización y mantenibilidad

    :param texto: Texto donde buscar
    :param query: Término de búsqueda
    :param opciones: Configuración
    :returns: Resultados
    """
    if opciones is None:
        opciones = {}

    config = {**SEARCH_CONFIG, **opciones}

    if not query or not texto:
        return {
            'matches': [],
            'total': 0,
            'positions': [],
            'highlighted': texto or '',
            'stats': {'duration': 0},
        }

    start_time = time.perf_counter()

    try:
        pattern = build_pattern(query, config)
        matches = []
        positions = []

        # Usar finditer para obtener toda la info (equivalente a matchAll)
        for match in pattern.finditer(texto):
            matches.append(match.group())

            if config['show_indices']:
                # Python siempre tiene .start()/.end() — no requiere flag 'd'
                positions.append({
                    'start': match.start(),
                    'end': match.end(),
                    'value': match.group(),
                })
            else:
                positions.append({
                    'start': match.start(),
                    'end': match.end(),
                    'value': match.group(),
                })

        highlighted = highlight(texto, pattern)
        duration = time.perf_counter() - start_time

        return {
            'matches': matches,
            'total': len(matches),
            'positions': positions,
            'highlighted': highlighted,
            'stats': {
                'duration': f'{duration * 1000:.2f}ms',
                'text_length': len(texto),
                'pattern': pattern.pattern,
            },
        }
    except re.error as error:
        return {
            'matches': [],
            'total': 0,
            'positions': [],
            'highlighted': texto,
            'stats': {'error': str(error)},
        }


def highlight(texto: str, pattern: re.Pattern) -> str:
    """
    Resaltar coincidencias en el texto

    :param texto: Texto original
    :param pattern: Patrón de búsqueda
    :returns: Texto con marcadores
    """
    return pattern.sub(r'[[MATCH]]\g<0>[[/MATCH]]', texto)


def highlight_to_html(texto: str) -> str:
    """
    Convertir marcadores a HTML
    """
    return (
        texto
        .replace('[[MATCH]]', '<mark class="highlight">')
        .replace('[[/MATCH]]', '</mark>')
    )


# ============================================
# BÚSQUEDA AVANZADA
# ============================================

def parse_query(query_string: str) -> dict:
    """
    Parsear query con sintaxis avanzada

    Soporta:
    - "frase exacta"
    - palabra1 palabra2 (AND implícito)
    - palabra1 OR palabra2
    - -excluir
    - palabra* (wildcard)

    :param query_string: Query del usuario
    :returns: Componentes parseados
    """
    parts = {
        'exact': [],
        'include': [],
        'exclude': [],
        'wildcard': [],
        'or': [],
    }

    remaining = query_string

    # 1. Extraer frases exactas "..."
    exact_pattern = re.compile(r'"([^"]+)"')
    for match in exact_pattern.finditer(remaining):
        parts['exact'].append(match.group(1))
    remaining = exact_pattern.sub(' ', remaining)

    # 2. Extraer exclusiones -palabra
    exclude_pattern = re.compile(r'-(\w+)')
    for match in exclude_pattern.finditer(remaining):
        parts['exclude'].append(match.group(1))
    remaining = exclude_pattern.sub(' ', remaining)

    # 3. Extraer wildcards palabra*
    wildcard_pattern = re.compile(r'(\w+)\*')
    for match in wildcard_pattern.finditer(remaining):
        parts['wildcard'].append(match.group(1))
    remaining = wildcard_pattern.sub(' ', remaining)

    # 4. Procesar OR
    or_pattern = re.compile(r'(\w+)\s+OR\s+(\w+)', re.IGNORECASE)
    for match in or_pattern.finditer(remaining):
        parts['or'].append([match.group(1), match.group(2)])
    remaining = or_pattern.sub(' ', remaining)

    # 5. El resto son palabras a incluir (AND)
    words = [w for w in remaining.strip().split() if w]
    parts['include'].extend(words)

    return parts


def busqueda_avanzada(texto: str, query_string: str, opciones: dict | None = None) -> dict:
    """
    Búsqueda avanzada con sintaxis especial

    :param texto: Texto donde buscar
    :param query_string: Query con sintaxis
    :param opciones: Configuración base
    :returns: Resultados
    """
    if opciones is None:
        opciones = {}

    config = {**SEARCH_CONFIG, **opciones}
    parsed = parse_query(query_string)
    resultados = []
    texto_filtrado = texto

    # Verificar exclusiones primero
    for excluir in parsed['exclude']:
        pattern = re.compile(rf'\b{escape_regex(excluir)}\b', re.IGNORECASE)
        if pattern.search(texto):
            # Si contiene palabra excluida, podría filtrar líneas
            lineas = texto.splitlines()
            texto_filtrado = '\n'.join(
                linea for linea in lineas if not pattern.search(linea)
            )

    # Buscar frases exactas
    for frase in parsed['exact']:
        resultado = buscar(texto_filtrado, frase, config)
        if resultado['total'] > 0:
            resultados.append({
                'type': 'exact',
                'query': frase,
                **resultado,
            })

    # Buscar wildcards
    for base in parsed['wildcard']:
        pattern = re.compile(rf'\b{escape_regex(base)}\w*\b', re.IGNORECASE)
        found = pattern.findall(texto_filtrado)
        resultados.append({
            'type': 'wildcard',
            'query': base + '*',
            'matches': list(set(found)),
            'total': len(found),
        })

    # Buscar OR
    for a, b in parsed['or']:
        pattern = re.compile(
            rf'\b(?:{escape_regex(a)}|{escape_regex(b)})\b',
            re.IGNORECASE,
        )
        found = pattern.findall(texto_filtrado)
        resultados.append({
            'type': 'or',
            'query': f'{a} OR {b}',
            'matches': found,
            'total': len(found),
        })

    # Buscar palabras individuales (AND)
    for palabra in parsed['include']:
        resultado = buscar(texto_filtrado, palabra, {
            **config,
            'whole_word': True,
        })
        if resultado['total'] > 0:
            resultados.append({
                'type': 'include',
                'query': palabra,
                **resultado,
            })

    # Verificar que todas las palabras include estén presentes (AND)
    todas_presentes = all(
        re.compile(rf'\b{escape_regex(palabra)}\b', re.IGNORECASE).search(texto_filtrado)
        for palabra in parsed['include']
    )

    return {
        'parsed': parsed,
        'resultados': resultados,
        'todas_presentes': todas_presentes,
        'texto_filtrado': texto_filtrado,
    }


# ============================================
# UTILIDADES ADICIONALES
# ============================================

def get_contexto(texto: str, posicion: int, caracteres: int = 50) -> str:
    """
    Obtener contexto alrededor de un match

    :param texto: Texto completo
    :param posicion: Posición del match
    :param caracteres: Caracteres de contexto
    :returns: Fragmento con contexto
    """
    start = max(0, posicion - caracteres)
    end = min(len(texto), posicion + caracteres)

    contexto = texto[start:end]

    if start > 0:
        contexto = '...' + contexto
    if end < len(texto):
        contexto = contexto + '...'

    return contexto


def get_lineas_contexto(texto: str, posicion: int, lineas: int = 2) -> dict:
    """
    Obtener líneas alrededor de un match

    :param texto: Texto completo
    :param posicion: Posición del match
    :param lineas: Líneas de contexto
    :returns: Líneas con contexto
    """
    todas_lineas = texto.splitlines()

    # Encontrar número de línea
    char_count = 0
    linea_actual = 0

    for i, linea in enumerate(todas_lineas):
        char_count += len(linea) + 1  # +1 por \n
        if char_count > posicion:
            linea_actual = i
            break

    inicio = max(0, linea_actual - lineas)
    fin = min(len(todas_lineas), linea_actual + lineas + 1)

    return {
        'lineas': todas_lineas[inicio:fin],
        'numero_linea_match': linea_actual,
        'rango': [inicio, fin],
    }


# ============================================
# TESTS
# ============================================

texto_ejemplo = """JavaScript es un lenguaje de programación.
JavaScript se usa para desarrollo web.
También se usa en Node.js para backend.
Python es otro lenguaje popular.
javascript puede ejecutarse en el navegador.
Los frameworks como React usan JavaScript."""

print('=== Búsqueda Simple ===\n')

resultado1 = buscar(texto_ejemplo, 'JavaScript')
print(f"Buscar 'JavaScript': {resultado1['total']} coincidencias")
print(f"Matches: {resultado1['matches']}")

print('\n=== Búsqueda Case-Sensitive ===\n')

resultado2 = buscar(texto_ejemplo, 'JavaScript', {'case_sensitive': True})
print(f"Buscar 'JavaScript' (case-sensitive): {resultado2['total']}")

resultado3 = buscar(texto_ejemplo, 'javascript', {'case_sensitive': True})
print(f"Buscar 'javascript' (case-sensitive): {resultado3['total']}")

print('\n=== Búsqueda Palabra Completa ===\n')

resultado4 = buscar(texto_ejemplo, 'Java', {'whole_word': True})
print(f"Buscar 'Java' (whole word): {resultado4['total']}")

resultado5 = buscar(texto_ejemplo, 'Java', {'whole_word': False})
print(f"Buscar 'Java' (parcial): {resultado5['total']}")

print('\n=== Búsqueda Avanzada ===\n')

query1 = '"desarrollo web" JavaScript -Python'
print(f'Query: {query1}')
print(busqueda_avanzada(texto_ejemplo, query1))

print('\n=== Contexto de Matches ===\n')

resultado6 = buscar(texto_ejemplo, 'Node.js')
if resultado6['positions']:
    pos = resultado6['positions'][0]['start']
    print(f"Contexto: {get_contexto(texto_ejemplo, pos, 30)}")
    print(f"Líneas: {get_lineas_contexto(texto_ejemplo, pos, 1)}")

print('\n=== Unicode ===\n')

texto_unicode = 'Hola 👋 mundo! Привет мир! 你好世界!'
resultado7 = buscar(texto_unicode, 'мир', {'unicode': True})
print(f"Buscar 'мир' en texto Unicode: {resultado7}")
